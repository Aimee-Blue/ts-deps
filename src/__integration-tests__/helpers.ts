import { spawnSync } from 'child_process'
import {
  readFile,
  writeFile,
  remove,
  emptyDir,
  pathExists,
  unlink,
} from 'fs-extra'
import { resolve, join, normalize, relative, isAbsolute } from 'path'
import { format, resolveConfig } from 'prettier'

export const PKG_JSON = 'package.json'
export const ROOT = resolve(__dirname, '../../')

// - start of the archive file name after yarn pack
const ARCH_PKG_FILE_NAME = 'zaripych-ts-deps'

export const patchPackageJsonVersion = async () => {
  console.log('Modifying package json')

  const packageJsonLocation = join(ROOT, PKG_JSON)

  const packageJsonContents = await readFile(packageJsonLocation, {
    encoding: 'utf-8',
  })

  const pkg = JSON.parse(packageJsonContents) as {
    version: string
  }

  if (!/-integration-test$/.test(pkg.version)) {
    pkg.version = `${pkg.version}-integration-test`

    const config = await resolveConfig(packageJsonLocation)

    await writeFile(
      packageJsonLocation,
      format(JSON.stringify(pkg), { ...config, parser: 'json' }),
      {
        encoding: 'utf-8',
      }
    )
  }

  return pkg.version
}

export const spawnAndCheck = (
  ...args: Parameters<typeof spawnSync>
): ReturnType<typeof spawnSync> => {
  const [app, cmdOrOpts, optsOrCb] = args

  const cmd = Array.isArray(cmdOrOpts) ? cmdOrOpts : []
  const opts = Array.isArray(cmdOrOpts) ? optsOrCb : cmdOrOpts

  const effectiveOpts = {
    ...opts,
    shell: true,
    encoding: 'utf-8',
  }

  console.log(app, Array.isArray(cmd) ? cmd.join(' ') : '')

  const result = spawnSync(app, cmd, effectiveOpts)

  console.log(result.output.filter(Boolean).join(''))

  if (result.error) {
    throw result.error
  }

  if (result.status != 0) {
    throw new Error(
      `process ${args[0]} exit with non-zero status code: ${result.status}`
    )
  }

  return result
}

const checkSafety = async (dir: string) => {
  const normalized = normalize(dir)
  const relativePath = relative(ROOT, normalized)

  if (/\.\./g.test(relativePath)) {
    throw new Error(
      'We can only delete directories within the root of our package'
    )
  }

  const exists = await pathExists(normalized)
  return { exists, path: normalized }
}

export const emptyDirSafe = async (relativeToRoot: string) => {
  const { exists, path } = await checkSafety(relativeToRoot)
  if (!exists) {
    return
  }

  await emptyDir(path)
}

export const rmDirSafe = async (relativeToRoot: string) => {
  const { exists, path } = await checkSafety(relativeToRoot)
  if (!exists) {
    return
  }

  await remove(path)
}

export const buildAndPack = async () => {
  const version = await patchPackageJsonVersion()

  console.log('version', version)

  spawnAndCheck('yarn', ['build'], {
    cwd: ROOT,
    shell: true,
  })

  const packageName = `${ARCH_PKG_FILE_NAME}-v${version}.tgz`
  const packageLocation = join(ROOT, packageName)

  await unlink(packageLocation).catch(() => Promise.resolve())

  spawnAndCheck('yarn', ['pack'], {
    cwd: ROOT,
    shell: true,
  })

  return { packageName, packageLocation, version }
}

const toUnixCompatiblePath = (path: string) => {
  if (!isAbsolute(path)) {
    throw new Error('Must be absolute path')
  }
  return '/' + path.replace(/\\/g, '/').replace(':', '')
}

export const unarchiveTarGz = async (cwd: string, tar: string, out: string) => {
  spawnAndCheck(
    'tar',
    ['zxvf', tar, '--directory', toUnixCompatiblePath(out)],
    {
      encoding: 'utf-8',
      shell: true,
      cwd,
    }
  )
}

const compareStrings = (a: string, b: string) => (a === b ? 0 : a > b ? 1 : -1)

const comparePathComponents = (a: string[], b: string[]): 0 | 1 | -1 => {
  if (a.length === 0 && b.length == 0) {
    return 0
  }
  const i = compareStrings(a[0], b[0])
  if (i === 0) {
    return comparePathComponents(a.slice(1), b.slice(1))
  } else {
    return i
  }
}

const seps = /\\|\//g

const comparePaths = (a: string, b: string) => {
  const componentsA = a.split(seps)
  const componentsB = b.split(seps)
  const result = comparePathComponents(componentsA, componentsB)
  return result
}

export const sortPaths = (files: string[]) => {
  files.sort(comparePaths)
}
