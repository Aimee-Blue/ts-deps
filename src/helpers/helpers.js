// @ts-check
import { resolve, basename } from 'path';
import { existsSync } from 'fs-extra';
import { prompt } from 'inquirer';

/**
 * @param {string} str
 */
function escapeRegExp(str) {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/**
 * @param {string} str
 * @param {string} character
 */
function trimLeft(str, character) {
  let charPrefixLength = 0;

  for (let i = 0; i < str.length; i++) {
    const char = str[i];

    if (char !== character) {
      charPrefixLength = i;
      break;
    }
  }

  return charPrefixLength ? str.substr(charPrefixLength) : str;
}

/**
 * @param {string} str
 * @param {string} character
 */
function trimRight(str, character) {
  let charSuffixLength = 0;

  for (let i = str.length - 1; i >= 0; i--) {
    const char = str[i];

    if (char !== character) {
      break;
    } else {
      charSuffixLength += 1;
    }
  }

  return charSuffixLength ? str.substr(0, str.length - charSuffixLength) : str;
}

/**
 * @param {string} str
 * @param {string} character
 */
function trim(str, character) {
  return trimRight(trimLeft(str, character), character);
}

/**
 * @param {string} str
 * @param {string} preferred
 */
function ensureSamePathSeparator(str, preferred = '/') {
  const regex = /\\|\//g;
  const result = str.replace(regex, preferred);
  return result;
}

/**
 * @param {string} p
 * @param {string} sep
 */
function ensureStartsWithPathSeparator(p, sep = '/') {
  const tested = ensureSamePathSeparator(p, sep);
  return tested.startsWith(sep) ? tested : `${sep}${tested}`;
}

/**
 * @param {string} p
 * @param {string} sep
 */
function ensureEndsWithPathSeparator(p, sep = '/') {
  const tested = ensureSamePathSeparator(p, sep);
  return tested.endsWith(sep) ? tested : `${tested}${sep}`;
}

/**
 * @param {string} p
 * @param {string} sep
 */
function trimPathSeparator(p, sep = '/') {
  const tested = ensureSamePathSeparator(p, sep);

  return trim(tested, sep);
}

/**
 * @param {string} p
 * @param {string} sep
 */
function ensureSurroundedWithPathSeparator(p, sep = '/') {
  const tested = ensureSamePathSeparator(p, sep);

  return `${sep}${trim(tested, sep)}${sep}`;
}

/**
 * @param {string[]} [candidates]
 */
function resolveTemplatesDir(candidates) {
  const templatesCandidates = candidates || [
    resolve(__dirname, '../../template-for-libs'),
  ];

  const templatesDir = templatesCandidates.find((cnd) => existsSync(cnd));
  if (!templatesDir) {
    throw new Error('Cannot find templates directory');
  }

  return templatesDir;
}

const copyPromptState = {
  overwriteAll: false,
};

/**
 * @param {{ overwrite: boolean | 'a' | 'c' }} result
 */
const handleCopyPromptResult = (result) => {
  if (result.overwrite === 'a') {
    copyPromptState.overwriteAll = true;
    return Promise.resolve(true);
  }

  if (result.overwrite === 'c') {
    process.exit(0);
  }

  if (typeof result.overwrite !== 'boolean') {
    throw new Error('Expected boolean as result of prompt');
  }

  return Promise.resolve(result.overwrite);
};

/**
 * @param {string} dest
 */
const promptForOverwrite = async (dest) => {
  if (copyPromptState.overwriteAll) {
    return Promise.resolve(true);
  }

  const fileName = basename(dest);

  const result = await prompt([
    {
      name: 'overwrite',
      message: `destination file "${fileName}" already exists, overwrite?`,
      type: 'expand',
      choices: [
        { name: 'Overwrite', value: true, key: 'y' },
        { name: 'Skip', value: false, key: 'n' },
        { name: 'Overwrite all files', value: 'a', key: 'a' },
        {
          name: 'Cancel (or just press Ctrl + C)',
          value: 'c',
          key: 'c',
        },
      ],
      default: true,
    },
  ]);

  return handleCopyPromptResult(result);
};
promptForOverwrite.state = copyPromptState;

export {
  trim,
  trimLeft,
  trimRight,
  escapeRegExp,
  trimPathSeparator,
  ensureSamePathSeparator,
  ensureStartsWithPathSeparator,
  ensureEndsWithPathSeparator,
  ensureSurroundedWithPathSeparator,
  resolveTemplatesDir,
  promptForOverwrite,
};
