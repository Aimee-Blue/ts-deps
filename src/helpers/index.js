export * from './helpers';
export * from './pullTemplate';
