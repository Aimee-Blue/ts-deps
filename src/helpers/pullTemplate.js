// @ts-check
const { spawnSync } = require('child_process');
const { unarchiveTarGz } = require('./unarchive');
const { resolveTemplatesDir } = require('./helpers');
const { join, relative, isAbsolute } = require('path');
const { pathExists, unlink, stat } = require('fs-extra');

const templateDirectories = ['template', 'template-max'];

/**
 * @param {string} depName
 * @param {string} cwd
 */
const npmPullTemplate = async (depName, cwd) => {
  const infoProcessResult = spawnSync('npm', ['info', depName], {
    encoding: 'utf8',
    shell: process.platform === 'win32',
    cwd,
  });

  if (infoProcessResult.status !== 0) {
    throw new Error(`Couldn't find '${depName}' package`);
  }

  const templatePathExist = await Promise.all(
    templateDirectories.map(dir => pathExists(join(cwd, dir)))
  );
  const existingTemplatePath = templatePathExist.indexOf(true);
  if (existingTemplatePath >= 0) {
    throw new Error(
      `Path already exists: ${
        templateDirectories[existingTemplatePath]
      }, it would be unsafe to pull templates to this location, please remove or backup the directory manually`
    );
  }

  const packResult = spawnSync('npm', ['pack', depName], {
    encoding: 'utf8',
    shell: process.platform === 'win32',
    cwd,
  });

  const packagePath = packResult.stdout.trim();
  const fullPackagePath = join(cwd, packagePath);

  await unarchiveTarGz(fullPackagePath, cwd, {
    map: header => {
      const name = header.name.replace('package/', '');
      return {
        ...header,
        name,
      };
    },
    ignore: (_name, headers) => {
      if (!headers) {
        return false;
      }
      return templateDirectories.every(dir => !headers.name.startsWith(dir));
    },
  });

  await unlink(fullPackagePath);

  return relative(
    cwd,
    resolveTemplatesDir(templateDirectories.map(dir => join(cwd, dir)))
  );
};

/**
 * @param {string} template
 * @param {string} cwd
 * @returns {Promise<TemplateInfo>}
 */
const pullTemplates = async (template, cwd) => {
  const templatePath = isAbsolute(template) ? template : join(cwd, template);
  const templatesStat = await stat(templatePath).catch(_err => null);

  if (templatesStat && templatesStat.isDirectory()) {
    return {
      type: 'directory',
      dir: templatePath,
    };
  }

  const result = await npmPullTemplate(template, cwd);

  return {
    type: 'package',
    dir: result,
  };
};

/**
 * @param {string} [template]
 * @param {string} cwd
 * @returns {Promise<TemplateInfo[]>}
 */
const initializeTemplates = async (template, cwd) => {
  /**
   * @type TemplateInfo
   */
  const defTemplateInfo = {
    type: 'default',
    dir: resolveTemplatesDir(),
  };

  const templateInfo = template
    ? await pullTemplates(template, cwd)
    : defTemplateInfo;

  const templates =
    templateInfo.type !== 'default'
      ? [defTemplateInfo, templateInfo]
      : [defTemplateInfo];

  return templates;
};

module.exports = {
  npmPullTemplate,
  initializeTemplates,
};
