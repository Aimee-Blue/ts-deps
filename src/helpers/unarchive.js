// @ts-check
import { mkdir, createReadStream } from 'fs-extra';
import { createGunzip } from 'zlib';
import { extract } from 'tar-fs';

/**
 * @param {string} tar
 * @param {string} out
 * @param {import('tar-fs').ExtractOptions} [opts]
 */
export const unarchiveTarGz = async (tar, out, opts) => {
  const outPath = out;
  await mkdir(outPath).catch(() => Promise.resolve());
  const gunzip = createGunzip();
  const stream = createReadStream(tar)
    .pipe(gunzip)
    .pipe(extract(outPath, opts), { end: true });
  return new Promise((res, rej) => {
    stream.once('end', () => {
      res();
    });
    stream.once('close', () => {
      res();
    });
    stream.once('finish', () => {
      res();
    });
    stream.once('error', err => {
      rej(err);
    });
  });
};
