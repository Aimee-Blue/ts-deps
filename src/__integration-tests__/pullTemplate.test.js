// @ts-check
import { ROOT, emptyDirSafe, sortedDirectoryContents } from './helpers';
import { join } from 'path';
import { ensureDir } from 'fs-extra';
import { npmPullTemplate } from '../helpers';

describe('npmPullTemplate', () => {
  describe('given fresh directory', () => {
    const pullDir = join(ROOT, 'integration-test-pull-template');

    beforeAll(async () => {
      await emptyDirSafe(pullDir);
      await ensureDir(pullDir);
    });

    it('should output correct files', async () => {
      const template = await npmPullTemplate('ts-deps@0.2.5', pullDir, pullDir);

      // for backward compatibility
      expect(template).toBe('template-max');

      const contents = await sortedDirectoryContents(join(pullDir, template));

      expect(contents).toMatchSnapshot();
    });
  });
});
