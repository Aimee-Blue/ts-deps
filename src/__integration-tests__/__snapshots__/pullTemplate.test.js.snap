// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`npmPullTemplate given fresh directory should output correct files 1`] = `
Array [
  "to-copy/",
  "to-copy/.yarnrc",
  "to-copy/babel.config.js",
  "to-copy/commitlint.config.js",
  "to-copy/jest.config.integration.js",
  "to-copy/jest.config.js",
  "to-copy/prettier.config.js",
  "to-copy/release.config.js",
  "to-copy/scripts/",
  "to-copy/scripts/build.js",
  "to-copy/scripts/clean.js",
  "to-copy/scripts/combineCoverage.js",
  "to-copy/tslint.json",
  "to-process/",
  "to-process/gitignore",
  "to-process/package.json",
]
`;
